idseq_dag/util/s3.py

```diff
@@ -217,53 +291,65 @@ def fetch_from_s3(src,  # pylint: disable=dangerous-default-value
                         log.write(f"Waited {wait_duration} seconds to acquire S3MI semaphore for {src}.")
 
                 if untar:
-                    write_dst = r''' | tar xvf - -C "${destdir}";'''
-                    named_args = {'destdir': destdir}
+                    write_dst = r''' | tar xvf - -C "${tmp_destdir}";'''
+                    named_args = {'tmp_destdir': tmp_destdir}
                 else:
-                    write_dst = r''' > "${dst}";'''
-                    named_args = {'dst': dst}
+                    write_dst = r''' > "${tmp_dst}";'''
+                    named_args = {'tmp_dst': tmp_dst}
                 command_params = f"{unzip} {write_dst}"
 
                 named_args.update({'src': src})
-                try:
-                    assert allow_s3mi
-                    command.execute(
-                        command_patterns.ShellScriptCommand(
-                            script=r'set -o pipefail; s3mi cat "${src}" ' + command_params,
-                            named_args=named_args
+
+                if os.path.exists(tmp_dst):
+                    command.remove_rf(tmp_dst)
+
+                try_cli = not allow_s3mi
+                if allow_s3mi:
+                    try:
+                        command.execute(
+                            command_patterns.ShellScriptCommand(
+                                script=r'set -o pipefail; s3mi cat --quiet "${src}" ' + command_params,
+                                named_args=named_args
+                            )
                         )
-                    )
-                except:
-                    if allow_s3mi:
+                    except:
+                        try_cli = not okay_if_missing
                         allow_s3mi = False
                         S3MI_SEM.release()
-                        log.write(
-                            "Failed to download with s3mi. Trying with aws s3 cp..."
-                        )
+                        if try_cli:
+                            log.write(
+                                "Failed to download with s3mi. Trying with aws s3 cp..."
+                            )
```

Comment:
We subsequently (a  month later) discovered a bug here, we are missing the lines
```
    else:
        raise
```
The impact is, in the event of s3mi I/O error, and when okay_if_missing=True, we do not wish to swallow the exception here because that causes "return dst" to be executed, and an incomplete or corrupt download to be used.

Even after fixing this, the design also has another trade off when okay_is_missing=True.   In case of a s3mi error, we presume the file is missing, instead of making double sure that's the case by doing try_cli anyway.  This was actually intentional here, since we were trying to reduce the number of excess aws s3 calls, we had too many of those, that were causing us problems.  The issue there is largely contained, however, so we could prefer to again be more cautious and always try_cli after s3mi fails.